src/heifimage.cpp

```diff
@@ -0,0 +1,390 @@
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, 5th Floor, Boston, MA 02110-1301 USA.
+ */
+/*!
+  @file    heifimage_int.hpp
+  @brief   Include HEIF header files.
+  @author  Peter Kovář (1div0)
+           <a href="mailto:peter.kovar@reflexion.tv">peter.kovar@reflexion.tv</a>
+  @date    25-Dec-18, 1div0: created
+  @author  Gilles Caulier (cgilles)
+           <a href="mailto:caulier dot gilles at gmail dot com">caulier dot gilles at gmail dot com</a>
+  @date    12-Oct-19, cgilles: add read exif, iptc, and xmp support
+ */
+// *****************************************************************************
+// included header files
+#include "config.h"
+
+#include "basicio.hpp"
+#include "error.hpp"
+#include "enforce.hpp"
+#include "futils.hpp"
+#include "image.hpp"
+#include "image_int.hpp"
+#include "heifimage_int.hpp"
+#include "types.hpp"
+
+// + standard includes
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <iterator>
+#include <string>
+
+// libheif includes
+#include <libheif/heif.h>
+
+const unsigned char HeifSignature[] = { 0x00, 0x00, 0x00, 0x18, 0x66, 0x74, 0x79, 0x70, 0x68, 0x65, 0x69, 0x63 };
+const unsigned char HeifBlank[]     = { 0x00, 0x00, 0x00, 0x18, 0x66, 0x74, 0x79, 0x70, 0x68, 0x65, 0x69, 0x63 };
+
+// *****************************************************************************
+// class member definitions
+namespace Exiv2 {
+    using namespace Exiv2::Internal;
+
+    HeifImage::HeifImage(BasicIo::UniquePtr io, bool create)
+            : Image(ImageType::heif, mdExif | mdIptc | mdXmp, std::move(io))
+    {
+        if (create)
+        {
+            if (io_->open() == 0)
+            {
+#ifdef EXIV2_DEBUG_MESSAGES
+                std::cerr << "Exiv2::HeifImage:: Creating HEIF image to memory" << std::endl;
+#endif
+                IoCloser closer(*io_);
+
+                if (io_->write(HeifBlank, sizeof(HeifBlank)) != sizeof(HeifBlank))
+                {
+#ifdef EXIV2_DEBUG_MESSAGES
+                    std::cerr << "Exiv2::HeifImage:: Failed to create HEIF image on memory" << std::endl;
+#endif
+                }
+            }
+        }
+    } // HeifImage::HeifImage
+
+    HeifImage::~HeifImage()
+    {
+    }
+
+    std::string HeifImage::mimeType() const
+    {
+        return "image/heif";
+    }
+
+    void HeifImage::setComment(const std::string& /*comment*/)
+    {
+        throw(Error(kerInvalidSettingForImage, "Image comment", "HEIF"));
+    } // HeifImage::setComment
+
+    void HeifImage::readMetadata()
```

Comment:
[suggestion] This function has almost 200 LoC. Would it be possible to extract some blocks of code into separated private methods or functions?